strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/ArtifactGroupServiceImpl.java

```diff
@@ -0,0 +1,104 @@
+package org.carlspring.strongbox.services.impl;
+
+import org.carlspring.strongbox.data.service.CommonCrudService;
+import org.carlspring.strongbox.domain.ArtifactGroup;
+import org.carlspring.strongbox.services.ArtifactGroupService;
+
+import javax.inject.Inject;
+import java.lang.reflect.UndeclaredThrowableException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.core.IMap;
+import com.orientechnologies.orient.core.sql.query.OSQLSynchQuery;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+/**
+ * @author Przemyslaw Fusik
+ */
+@Service
+@Transactional
+public class ArtifactGroupServiceImpl
+        extends CommonCrudService<ArtifactGroup>
+        implements ArtifactGroupService
+{
+
+    private static final String ARTIFACT_GROUP_MAP_LOCKS = "artifact-group-map-locks";
+
+    @Inject
+    private HazelcastInstance hazelcastInstance;
+
+    @Override
+    public Class<ArtifactGroup> getEntityClass()
+    {
+        return ArtifactGroup.class;
+    }
+
+    @Override
+    public <T extends ArtifactGroup> T findOneOrCreate(Class<T> type,
+                                                       String name)
+    {
+        Map<String, String> params = new HashMap<>();
+        params.put("name", name);
+        params.put("type", type.getName());
+
+        String sQuery = buildQuery(params);
+
+        OSQLSynchQuery<Long> oQuery = new OSQLSynchQuery<>(sQuery);
+        oQuery.setLimit(1);
+
+        // make sure there are no duplicates per `params`
+        lock(params);
```

Comment:
I think we can use `NeedRetryException` here instead of locking.

Reply:
yeah, thanks :)